altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+
+        // Regardless of block header size, there should be more than 0 memory remaining
+        match block_hdr_size {
+            16 => assert_eq!(remaining_size, 320 - (128 + 64 + 64)),
+            8 => assert_eq!(remaining_size, 328 - (128 + 64 + 56)),
+            _ => assert!(remaining_size > 0),
+        }
+    }
+
+    #[test]
+    fn deallocations_do_not_leak_memory() {
+        // Should get aligned down based on block header size
+        // Should be 328 if block header size is 8, 320 if it's 16
+        let heap_size: usize = 330;
+        let block_hdr_size = mem::size_of::<BlockHeader>();
+        let aligned_heap_size = alignment::align_down(heap_size, block_hdr_size);
+
+        let mut tfl = test::get_free_list_with_size(heap_size);
+
+        let alloc1 = tfl.free_list.allocate(60, 1);
+        let alloc2 = tfl.free_list.allocate(122, 1);
+        let alloc3 = tfl.free_list.allocate(54, 1);
+        tfl.free_list.deallocate(alloc1, 60, 1);
+        tfl.free_list.deallocate(alloc2, 122, 1);
+        tfl.free_list.deallocate(alloc3, 54, 1);
+
+        assert_eq!(tfl.sum_free_block_memory(), aligned_heap_size);
+        assert!(
+            tfl.each_free_block_satisfies(
+                &|current| unsafe { (*current).block_size % block_hdr_size == 0 }
+            )
+        );
+    }
+
+    #[test]
+    fn many_allocations_and_deallocations_do_not_leak_memory() {
+        // Should get aligned down based on block header size
+        // Should be 328 if block header size is 8, 320 if it's 16
+        let heap_size: usize = 330;
+        let block_hdr_size = mem::size_of::<BlockHeader>();
+        let aligned_heap_size = alignment::align_down(heap_size, block_hdr_size);
+
+        let mut tfl = test::get_free_list_with_size(heap_size);
+
+        let alloc1 = tfl.free_list.allocate(60, 1);
+        let alloc2 = tfl.free_list.allocate(122, 1);
+        tfl.free_list.deallocate(alloc1, 60, 1);
+        let alloc3 = tfl.free_list.allocate(54, 4);
+        tfl.free_list.deallocate(alloc2, 122, 1);
+        let alloc4 = tfl.free_list.allocate(36, 8);
+        tfl.free_list.deallocate(alloc3, 54, 4);
+        tfl.free_list.deallocate(alloc4, 36, 8);
+        let alloc5 = tfl.free_list.allocate(8, 32);
+        let alloc6 = tfl.free_list.allocate(4, 1);
+        tfl.free_list.deallocate(alloc5, 8, 32);
+        tfl.free_list.deallocate(alloc6, 4, 1);
+
+        assert_eq!(tfl.sum_free_block_memory(), aligned_heap_size);
+        assert!(
+            tfl.each_free_block_satisfies(
+                &|current| unsafe { (*current).block_size % block_hdr_size == 0 }
+            )
+        );
+    }
+
+    #[test]
+    fn out_of_memory_returns_null() {
+        let heap_size: usize = 512;
+        let mut tfl = test::get_free_list_with_size(heap_size);
+
+        tfl.free_list.allocate(512, 1);
+        assert!(tfl.free_list.allocate(64, 1).is_null());
+    }
+
+    #[test]
+    fn not_enough_memory_returns_null() {
```

Comment:
In this case, the list has some memory remaining (256 bytes) but not enough to fulfill the request, while the previous case has no remaining memory, so the list should be empty. I should add a comment to clarify this.